### PR TITLE
Tab completer for existing resources (Iss #5415)

### PR DESCRIPTION
**Pull request by vladimir-shcherbakov**
## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by maddieclayton** [comment]
@vladimir-shcherbakov is this ready for review? If not, please add the "Do not merge" tag.

**Comment by vladimir-shcherbakov** [comment]
@maddieclayton  
Yes, it's ready.

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                var contextHash = HashContext(context, resourceType);
+                var cacheItem = Cache.ContainsKey(contextHash) ? Cache[contextHash] : null;
+
+                if (cacheItem != null && DateTime.Now.Subtract(cacheItem.Timestamp).CompareTo(TimeToUpdate) < 0 )
+                {
+                    return Cache[contextHash].ResourceInfoList;
+                }
+
+                var instance = AzureSession.Instance;
+                var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
```
Comment (maddieclayton) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: same comment here about createarmclient

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        {
+            return (context.Account.Id + context.Environment.Name + context.Subscription.Id + context.Tenant.Id + resourceType).GetHashCode();
+        }
+
+        public partial class ResourceInfo
+        {
+            private static class Patterns
+            {
+                // example: /subscriptions/1c638cf4-608f-4ee6-b680-c329e824c3a8/resourceGroups/rgrssdfw69c0219330/providers/Microsoft.Sql/servers/sqlserver2c5254480/databases/master
+                private const string SubscripitonId = @"([0-9a-f]{8}[-][0-9a-f]{4}[-][0-9a-f]{4}[-][0-9a-f]{4}[-][0-9a-f]{12})";
```
Comment (maddieclayton) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: subscription spelled wrong

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                private const string ServiceName = @"([^/]+)";
+                private const string ResourceName = @"([^/]+)";
+                private const string Sub = @"(/[^/]+/[^/]+)?";
+
+                public static readonly string ResourceId = $"/subscriptions/{SubscripitonId}/resourceGroups/{ResourceGroupName}/providers/{ProviderName}/{ServiceName}/{ResourceName}{Sub}";
+            }
+
+            public static ResourceInfo FromResourceId(string resourceId)
+            {
+                var match = Regex.Match(resourceId, Patterns.ResourceId, RegexOptions.IgnoreCase);
```
Comment (maddieclayton) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: The numbers here seem a little magically to me, but this is fine as long as you've tested it on resources at multiple levels.

File: src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.cs [hunk 4]
```diff
@@ -48,5 +48,19 @@ public void TestResourceGroupCompleter()
         {
             ProfileController.NewInstance.RunPsTest(xunitLogger, "72f988bf-86f1-41af-91ab-2d7cd011db47", "Test-ResourceGroupCompleter");
         }
+
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.Flaky)]
```
Comment (maddieclayton) [inline src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.cs]: Why did you add flaky?

File: src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1 [hunk 5]
```diff
@@ -39,4 +39,53 @@ function Test-ResourceGroupCompleter
...
+    # take data from Azure and put to cache
+    [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceNameCompleterAttribute]::new($resourceType, "ResourceGroupName", $Null)
+    $resourceNames = [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceNameCompleterAttribute]::GetResources($resourceType) | ForEach-Object { $_.Name }
+    Assert-AreEqualArray $resourceNames $expectResourceNames
+    # take data from the cache
+    $resourceNames = [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceNameCompleterAttribute]::GetResources($resourceType) | ForEach-Object { $_.Name }
+    Assert-AreEqualArray $resourceNames $expectResourceNames
+    # change time to update the cache
+    [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceNameCompleterAttribute]::TimeToUpdate = [System.TimeSpan]::FromSeconds(1)
+    Start-Sleep -Seconds 1
```
Comment (maddieclayton) [inline src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1]: Use wait-seconds: https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Common/Commands.ScenarioTests.ResourceManager.Common/Common.ps1#L307. We don't allow sleeps in tests.

File: src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1 [hunk 6]
```diff
@@ -39,4 +39,53 @@ function Test-ResourceGroupCompleter
...
+    $expectResourceIds = Get-AzureRmResource -ResourceType  $resourceType | ForEach-Object {$_.Id}
+    # take data from Azure and put to cache
+    $resourceIds = [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceIdCompleterAttribute]::GetResourceIds($resourceType)
+    Assert-AreEqualArray $resourceIds $expectResourceIds
+    # take data from the cache
+    $resourceIds = [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceIdCompleterAttribute]::GetResourceIds($resourceType)
+    Assert-AreEqualArray $resourceIds $expectResourceIds
+    # change time to update the cache
+    [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceIdCompleterAttribute]::TimeToUpdate = [System.TimeSpan]::FromSeconds(1)
+    Start-Sleep -Seconds 1
```
Comment (maddieclayton) [inline src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1]: same comment here.

**Comment by maddieclayton** [comment]
@vladimir-shcherbakov Looks like your test is failing on travis - please check

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                var contextHash = HashContext(context, resourceType);
+                var cacheItem = Cache.ContainsKey(contextHash) ? Cache[contextHash] : null;
+
+                if (cacheItem != null && DateTime.Now.Subtract(cacheItem.Timestamp).CompareTo(TimeToUpdate) < 0)
+                {
+                    return Cache[contextHash].ResourceInfoList;
+                }
+
+                var instance = AzureSession.Instance;
+                var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
```
[the remark this answers is not in the document]
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs]: fixed

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                    context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                    instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                    instance.ClientFactory.GetCustomHandlers());
+
+                client.SubscriptionId = context.Subscription.Id;
```
[the remark this answers is not in the document]
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs]: fixed

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+                var odata = new ODataQuery<GenericResourceFilter>(r => r.ResourceType == resourceType);
+
+                IEnumerable<string> resourceInfoList;
+                try
+                {
+                    resourceInfoList = client.Resources.List(odata)
```
[the remark this answers is not in the document]
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs]: fixed

File: src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.cs (hunk 4 above)
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.cs]: copied from other completes tests.
fixed

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            string resourceType,
+            string resourceGroupParameterName,
+            string nodeServiceParameterName)
+        {
+            var script = "param($commandName, $parameterName, $wordToComplete, $commandAst, $fakeBoundParameters)\n" +
+                $"$resourceType = \"{resourceType}\"\n" +
+                $"$resourceGroupParameterName = \"{resourceGroupParameterName ?? "$Null"}\"\n" +
+                $"$nodeServiceParameterName = \"{nodeServiceParameterName ?? "$Null"}\"\n" +
+                "$ids = [Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters.ResourceNameCompleterAttribute]::GetResources($resourceType)\n" +
+                "if ($resourceGroupParameterName -ne $Null -and $fakeBoundParameters.Contains($resourceGroupParameterName)) { $ids = $ids | Where-Object { $_.Group -eq $resourceGroupParameterName }}\n" +
```
[the remark this answers is not in the document]
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: In this context, if a user has provided a resource group name:
```powershell
Get-AzrureRmSomething -ResouceGroupName "grName" -SomethingName ...
```
I gets used as a filter for the next ```SomethingName``` argument.

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+                "if ($nodeServiceParameterName -ne $Null -and $fakeBoundParameters.Contains($nodeServiceParameterName)) { $ids = $ids | Where-Object { $_.NodeName -eq $nodeServiceParameterValue }}\n" +
```
[the remark this answers is not in the document]
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: This is the idea.
Let's say we have a cmdlet:
```Powershell
Get-AzureRmDatabase [ResourceGroupName <string>] [ServerName <string>] [DbName <string>]
```
and class with the ResourceIdCompleter
```Cs
public class GetAzureRmDatabaseCmdlet {
...
   [Paremeter]
   [ResourceIdCompleter("Microsoft.Sql/servers/databases")]
   public string DbName {get; set;}
...
}
```
and resource id template
```
/subscriptions/{sub-id}/resourceGroups/{rg-name}/providers/Microsoft.Sql/servers/{server-name}/databases/{db-name}
```
If a customer's input is like this: 
```Powershell
Get-AzureRmDatabase -DbName ...
```
the completer knows nothing about resource group name and server name, so it returns all db-names in Microsoft.Sql/servers provider space

If a customer's input is like this: 
```Powershell
Get-AzureRmDatabase -ResourceGroupName "rgName" -ServerName "servername" -DbName ...
```
the completer uses "rgName" and "servername" as filters.

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs (hunk 1 above)
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: fixed

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs (hunk 2 above)
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: fixed

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs (hunk 3 above)
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceNameCompleter.cs]: 2 levels only

File: src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1 (hunk 5 above)
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1]: fixed

File: src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1 (hunk 6 above)
Comment (vladimir-shcherbakov) [reply src/ResourceManager/Profile/Commands.Profile.Test/ArgumentCompleterTests.ps1]: fixed

**Comment by maddieclayton** [comment]
@vladimir-shcherbakov Can you remove ResourceNameCompleter?

**Comment by vladimir-shcherbakov** [comment]
ResourceNameCompleter removed

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                var cacheItem = Cache.ContainsKey(contextHash) ? Cache[contextHash] : null;
+
+                if (cacheItem != null && DateTime.Now.Subtract(cacheItem.Timestamp).CompareTo(TimeToUpdate) < 0)
+                {
+                    return Cache[contextHash].ResourceInfoList;
+                }
+
+                var client = AzureSession.Instance.ClientFactory.CreateArmClient<ResourceManagementClient>(context, AzureEnvironment.Endpoint.ResourceManager);
+
+                client.SubscriptionId = context.Subscription.Id;
```
Comment (maddieclayton) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceIdCompleter.cs]: Remove this, the "createarmclient" call does this automatically

**Comment by maddieclayton** [comment]
@vladimir-shcherbakov One small comment then LGTM.  Updated your PR to include the changelog fix, so should pass this time.

**Comment by praries880** [comment]
on demand : https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/939/

**Pull request merged**